### PR TITLE
[test] [bug] Use runpy in test_runtime.py to fix #1406

### LINKED ISSUE #1406: [Test] [bug] "test_runtime.py" failing on Windows
**Describe the bug**
`ti.exe` may not work properly on Windows in some cases -- `test_runtime.py::test_without_init` (which introduced in #1347) fails on my end because it calls `ti.exe`.

**Log/Screenshots**
```
tests/python/test_runtime.py::test_without_init [Taichi] mode=development
[Taichi] <dev mode>, llvm 8.0.1, commit f2bd9829, python 3.7.4

*******************************************
**      Taichi Programming Language      **
*******************************************
C:\Users\xmk\AppData\Local\Temp\tmpqrx9npc8 is not a valid command!
usage:
    baseline            |-> Archive current benchmark result as baseline
    benchmark           |-> Run Python tests in benchmark mode
    build               |-> Build C++ files
    changelog           |-> Display changelog of current version
    debug               |-> Debug a single script
    dist                |-> Build package and test in release mode
    doc                 |-> Build documentation
    example             |-> Run an example by name (or name.py)
    format              |-> Reformat modified source files
    format_all          |-> Reformat all source files
    gif                 |-> Convert mp4 file to gif in the same directory
    regression          |-> Display benchmark regression test result
    release             |-> Make source code release
    run                 |-> Run a single script
    task                |-> Run a specific task
    test                |-> Run the tests
    update              |-> Update the Taichi codebase
    video               |-> Make a video using *.png files in the current directory
    video_crop          |-> Crop video in the same directory
    video_scale         |-> Scale video resolution in the same directory
    video_speed         |-> Speed up video in the same directory

Taichi CLI

positional arguments:
  command     command from the above list to run

optional arguments:
  -h, --help  show this help message and exit
>>> Running time: 0.00s
FAILED

=========================================== FAILURES ===========================================
______________________________________ test_without_init _______________________________________

    def test_without_init():
        # We want to check if Taichi works well without ``ti.init()``.
        # But in test ``ti.init()`` will always be called in last ``@ti.all_archs``.
        # So we have to create a new Taichi instance, i.e. test in a sandbox.
        content = '''
    import taichi as ti
    assert ti.cfg.arch == ti.cpu
    
    x = ti.var(ti.i32, (2, 3))
    assert x.shape == (2, 3)
    
    x[1, 2] = 4
    assert x[1, 2] == 4
    '''
        filename = make_temp_file()
        with open(filename, 'w') as f:
            f.write(content)
>       assert os.system(f'{sys.executable} {filename}') == 0
E       assert 1 == 0
E         -1
E         +0

tests\python\test_runtime.py:23: AssertionError
============================== 1 failed, 3 passed in 3.59 seconds ==============================
>>> Running time: 4.21s

```

**To Reproduce**
The failing test:
```python
import taichi as ti
from taichi import make_temp_file
import sys, os


def test_without_init():
    # We want to check if Taichi works well without ``ti.init()``.
    # But in test ``ti.init()`` will always be called in last ``@ti.all_archs``.
    # So we have to create a new Taichi instance, i.e. test in a sandbox.
    content = '''
import taichi as ti
assert ti.cfg.arch == ti.cpu

x = ti.var(ti.i32, (2, 3))
assert x.shape == (2, 3)

x[1, 2] = 4
assert x[1, 2] == 4
'''
    filename = make_temp_file()
    with open(filename, 'w') as f:
        f.write(content)
    assert os.system(f'{sys.executable} {filename}') == 0
```

This works:
```python
import taichi as ti
from taichi import make_temp_file
import sys, os


def test_without_init():
    # We want to check if Taichi works well without ``ti.init()``.
    # But in test ``ti.init()`` will always be called in last ``@ti.all_archs``.
    # So we have to create a new Taichi instance, i.e. test in a sandbox.
    content = '''
import taichi as ti
assert ti.cfg.arch == ti.cpu

x = ti.var(ti.i32, (2, 3))
assert x.shape == (2, 3)

x[1, 2] = 4
assert x[1, 2] == 4
'''
    filename = make_temp_file()
    with open(filename, 'w') as f:
        f.write(content)
    assert os.system(f'python -m taichi {filename}') == 0
```
(Corresponding log:)
```
tests/python/test_runtime.py::test_without_init [Taichi] mode=development
[Taichi] <dev mode>, llvm 8.0.1, commit f2bd9829, python 3.7.4

*******************************************
**      Taichi Programming Language      **
*******************************************
C:\Users\xmk\AppData\Local\Temp\tmpr6112yd1 is not a valid command!
usage:
    baseline            |-> Archive current benchmark result as baseline
    benchmark           |-> Run Python tests in benchmark mode
    build               |-> Build C++ files
    changelog           |-> Display changelog of current version
    debug               |-> Debug a single script
    dist                |-> Build package and test in release mode
    doc                 |-> Build documentation
    example             |-> Run an example by name (or name.py)
    format              |-> Reformat modified source files
    format_all          |-> Reformat all source files
    gif                 |-> Convert mp4 file to gif in the same directory
    regression          |-> Display benchmark regression test result
    release             |-> Make source code release
    run                 |-> Run a single script
    task                |-> Run a specific task
    test                |-> Run the tests
    update              |-> Update the Taichi codebase
    video               |-> Make a video using *.png files in the current directory
    video_crop          |-> Crop video in the same directory
    video_scale         |-> Scale video resolution in the same directory
    video_speed         |-> Speed up video in the same directory

Taichi CLI

positional arguments:
  command     command from the above list to run

optional arguments:
  -h, --help  show this help message and exit
>>> Running time: 0.00s
PASSED
```

**If you have local commits (e.g. compile fixes before you reproduce the bug), please make sure you first make a PR to fix the build errors and then report the bug.**

### PR DESCRIPTION
**Pull request by archibate**
Related issue = close #1406

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
@xumingkuan Could you confirm the issue is gone?

### CODE REVIEW

**Comment by xumingkuan** [comment]
Thanks for trying to fix this, but the test fails on my end:
```
C:\Users\xmk\Desktop\taichi>ti test runtime -t 1
[Taichi] mode=development
[Taichi] <dev mode>, llvm 8.0.1, commit 49c4e030, python 3.7.4

*******************************************
**      Taichi Programming Language      **
*******************************************

Running Python tests...

Starting 1 testing thread(s)...
D:\Anaconda3\lib\site-packages\win32\lib\pywintypes.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative
uses
  import imp, sys, os
D:\Anaconda3\lib\site-packages\pyreadline\py3k_compat.py:8: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, an
d in 3.8 it will stop working
  return isinstance(x, collections.Callable)
============================================================================== test session starts ==============================================================================
platform win32 -- Python 3.7.4, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: C:\Users\xmk\Desktop\taichi
plugins: cov-2.9.0, forked-1.1.3, rerunfailures-9.0, xdist-1.32.0, arraydiff-0.3, doctestplus-0.3.0, openfiles-0.3.2, remotedata-0.3.2
collected 4 items                                                                                                                                                                

tests\python\test_runtime.py ...F                                                                                                                                          [100%]

=================================================================================== FAILURES ====================================================================================
_______________________________________________________________________________ test_without_init _______________________________________________________________________________

    def test_without_init():
        # We want to check if Taichi works well without ``ti.init()``.
        # But in test ``ti.init()`` will always be called in last ``@ti.all_archs``.
        # So we have to create a new Taichi instance, i.e. test in a sandbox.
        run_in_sandbox('''
    import taichi as ti
    assert ti.cfg.arch == ti.cpu
    
    x = ti.var(ti.i32, (2, 3))
    assert x.shape == (2, 3)
    
    x[1, 2] = 4
    assert x[1, 2] == 4
>   ''')

tests\python\test_runtime.py:20:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python\taichi\misc\test.py:41: in run_in_sandbox
    return runpy.run_path(filename)
D:\Anaconda3\lib\runpy.py:263: in run_path
    pkg_name=pkg_name, script_name=fname)
D:\Anaconda3\lib\runpy.py:96: in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
D:\Anaconda3\lib\runpy.py:85: in _run_code
    exec(code, run_globals)
..\..\AppData\Local\Temp\tmpahkamfjf:5: in <module>
    x = ti.var(ti.i32, (2, 3))
python\taichi\lang\util.py:212: in wrapped
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dt = DataType.int32, shape = (2, 3), offset = None, needs_grad = False

    @python_scope
    def var(dt, shape=None, offset=None, needs_grad=False):
        if isinstance(shape, numbers.Number):
            shape = (shape, )
    
        if isinstance(offset, numbers.Number):
            offset = (offset, )
    
        if shape is not None and offset is not None:
            assert len(shape) == len(
                offset
            ), f'The dimensionality of shape and offset must be the same  (f{len(shape)} != f{len(offset)})'
    
        assert (offset is not None and shape is None
                ) == False, f'The shape cannot be None when offset is being set'
    
        if get_runtime().materialized:
            raise RuntimeError(
>               "No new variables can be declared after materialization, i.e. kernel invocations "
                "or Python-scope tensor accesses. I.e., data layouts must be specified before "
                "any computation. Try appending ti.init() or ti.reset() "
                "right after 'import taichi as ti' if you are using Jupyter notebook."
            )
E           RuntimeError: No new variables can be declared after materialization, i.e. kernel invocations or Python-scope tensor accesses. I.e., data layouts must be specified be
fore any computation. Try appending ti.init() or ti.reset() right after 'import taichi as ti' if you are using Jupyter notebook.

python\taichi\lang\impl.py:297: RuntimeError
====================================================================== 1 failed, 3 passed in 2.40 seconds =======================================================================
>>> Running time: 2.94s
0.000000
```

**Comment by xumingkuan** [comment]
While
```python
run_in_sandbox('''
import taichi as ti
assert ti.cfg.arch == ti.cpu

x = ti.var(ti.i32, (2, 3))
assert x.shape == (2, 3)

x[1, 2] = 4
assert x[1, 2] == 4
''')
```
fails,
```python
run_in_sandbox('''
import taichi as ti
assert ti.cfg.arch == ti.cpu
ti.init()  # <---

x = ti.var(ti.i32, (2, 3))
assert x.shape == (2, 3)

x[1, 2] = 4
assert x[1, 2] == 4
''')
```
passes...

File: python/taichi/misc/test.py [hunk 1]
```diff
@@ -31,3 +31,10 @@ def make_temp_file(*args, **kwargs):
     fd, name = mkstemp(*args, **kwargs)
     os.close(fd)
     return name
+
+
+def run_in_sandbox(content):
```
Comment (xumingkuan) [inline python/taichi/misc/test.py:36]:
```suggestion
def run_in_sandbox(content):
    import runpy
```

File: tests/python/test_runtime.py [hunk 2]
```diff
@@ -1,13 +1,14 @@
 import taichi as ti
-from taichi import make_temp_file
+from taichi import run_in_sandbox
 import sys, os
+import runpy
```
Comment (xumingkuan) [inline tests/python/test_runtime.py:4]:
```suggestion
```

**Comment by xumingkuan** [comment]
FYI, the test runs successfully in my CLI:
```
(base) C:\Users\xmk>python
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import taichi as ti
[Taichi] mode=development
[Taichi] <dev mode>, llvm 8.0.1, commit 49c4e030, python 3.7.4
[Taichi] Interactive shell detected: Python shell
>>> x = ti.var(ti.i32, (2, 3))
[T 07/05/20 15:26:18.171] [program.cpp:taichi::lang::Program::Program@52] Program initializing...
[T 07/05/20 15:26:18.171] [memory_pool.cpp:taichi::lang::MemoryPool::MemoryPool@10] Memory pool created. Default buffer size per allocator = 1024 MB
[T 07/05/20 15:26:18.172] [llvm_context.cpp:taichi::lang::TaichiLLVMContext::TaichiLLVMContext@56] Creating Taichi llvm context for arch: x64
[T 07/05/20 15:26:18.172] [llvm_context.cpp:taichi::lang::TaichiLLVMContext::get_this_thread_data@684] Creating thread local data for thread 21240
[T 07/05/20 15:26:18.173] [llvm_context.cpp:taichi::lang::TaichiLLVMContext::TaichiLLVMContext@81] Taichi llvm context created.
[T 07/05/20 15:26:18.173] [program.cpp:taichi::lang::Program::Program@167] Program (0x1ea5a29cb60) arch=x64 initialized.
[D 07/05/20 15:26:18.174] [snode.cpp:taichi::lang::SNode::create_node@51] Non-power-of-two node size 3 promoted to 4.
>>> assert x.shape == (2, 3)
[T 07/05/20 15:26:24.804] [C:\Users\xmk\Desktop\taichi\python\taichi\lang\impl.py:materialize@181] Materializing layout...
[T 07/05/20 15:26:25.699] [llvm_context.cpp:taichi::lang::compile_runtime_bitcode@180] Compiling runtime module bitcode...
[T 07/05/20 15:26:26.149] [llvm_context.cpp:taichi::lang::compile_runtime_bitcode@193] Runtime module bitcode compiled.
[T 07/05/20 15:26:26.535] [unified_allocator.cpp:taichi::lang::UnifiedAllocator::UnifiedAllocator@53] Allocating virtual address space of size 1024 MB
[T 07/05/20 15:26:26.536] [unified_allocator.cpp:taichi::lang::UnifiedAllocator::UnifiedAllocator@62] Memory allocated. Allocation time = 0.00088 s
[T 07/05/20 15:26:26.537] [program.cpp:taichi::lang::Program::initialize_runtime_system@244] Allocating data structure of size 32 B
[T 07/05/20 15:26:26.539] [program.cpp:taichi::lang::Program::initialize_runtime_system@253] LLVMRuntime initialized
[T 07/05/20 15:26:26.540] [program.cpp:taichi::lang::Program::initialize_runtime_system@255] LLVMRuntime pointer fetched
[D 07/05/20 15:26:26.542] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@102] Processing memory alloc request 0
[D 07/05/20 15:26:26.542] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@109]   Allocating memory 8232 B (alignment 4096B)
[D 07/05/20 15:26:26.543] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@111]   Allocated. Ptr = 0x1ea003a9000
[D 07/05/20 15:26:26.545] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@102] Processing memory alloc request 1
[D 07/05/20 15:26:26.545] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@109]   Allocating memory 8232 B (alignment 4096B)
[D 07/05/20 15:26:26.549] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@111]   Allocated. Ptr = 0x1ea003ac000
[D 07/05/20 15:26:26.551] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@102] Processing memory alloc request 2
[D 07/05/20 15:26:26.551] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@109]   Allocating memory 8232 B (alignment 4096B)
[D 07/05/20 15:26:26.552] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@111]   Allocated. Ptr = 0x1ea003af000
[D 07/05/20 15:26:26.555] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@102] Processing memory alloc request 3
[D 07/05/20 15:26:26.555] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@109]   Allocating memory 3145728 B (alignment 4096B)
[D 07/05/20 15:26:26.559] [memory_pool.cpp:taichi::lang::MemoryPool::daemon@111]   Allocated. Ptr = 0x1ea003b2000
[T 07/05/20 15:26:26.559] [program.cpp:taichi::lang::Program::materialize_layout@321] materialize_layout called
>>> x[1, 2] = 4
>>> assert x[1, 2] == 4
>>> exit()
```

**Comment by archibate** [comment]
> While
> 
> ```python
> run_in_sandbox('''
> import taichi as ti
> assert ti.cfg.arch == ti.cpu
> 
> x = ti.var(ti.i32, (2, 3))
> assert x.shape == (2, 3)
> 
> x[1, 2] = 4
> assert x[1, 2] == 4
> ''')
> ```
> 
> fails,
> 
> ```python
> run_in_sandbox('''
> import taichi as ti
> assert ti.cfg.arch == ti.cpu
> ti.init()  # <---
> 
> x = ti.var(ti.i32, (2, 3))
> assert x.shape == (2, 3)
> 
> x[1, 2] = 4
> assert x[1, 2] == 4
> ''')
> ```
> 
> passes...

I just want to test if the code work well without ``ti.init()``..
If added, then this test have no effect at all... I think we should `fork` in to a new process and execute it so that `taichi` gets reloaded?

**Comment by xumingkuan** [comment]
> I think we should `fork` into a new process and execute it so that `taichi` gets reloaded?

Feel free to have a try.

**Comment by xumingkuan** [comment]
I just googled what `sys.executable` is, and realized that if I use `ti test runtime -t 1`, `sys.executable` will be `ti.exe` and then `ti.exe temp_file` fails.

Maybe I should just use `python -m taichi test runtime -t 1` as a workaround. `sys.executable` in this case is `python.exe` and it works.

**Comment by archibate** [comment]
> > I think we should `fork` into a new process and execute it so that `taichi` gets reloaded?
> 
> Feel free to have a try.

Is `os.fork` available on win?
I think the best idea is to `ti.exe` -> `ti.bat`...

**Comment by xumingkuan** [comment]
> > > I think we should `fork` into a new process and execute it so that `taichi` gets reloaded?
> > 
> > 
> > Feel free to have a try.
> 
> Is `os.fork` available on win?
> I think the best idea is to `ti.exe` -> `ti.bat`...

No...
```
(base) C:\Users\xmk\Desktop\taichi>python
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.fork
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'os' has no attribute 'fork'
>>> os.fork()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'os' has no attribute 'fork'
>>>
```

**Pull request closed without merging**